components/script/dom/bluetoothpermissionresult.rs

```diff
@@ -0,0 +1,219 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+use bluetooth_traits::{BluetoothRequest, BluetoothResponse};
+use bluetooth_traits::scanfilter::{BluetoothScanfilter, BluetoothScanfilterSequence};
+use dom::bindings::cell::DOMRefCell;
+use dom::bindings::codegen::Bindings::BluetoothPermissionResultBinding::{self, BluetoothPermissionResultMethods};
+use dom::bindings::codegen::Bindings::NavigatorBinding::NavigatorBinding::NavigatorMethods;
+use dom::bindings::codegen::Bindings::PermissionStatusBinding::{PermissionState, PermissionStatusMethods};
+use dom::bindings::codegen::Bindings::WindowBinding::WindowBinding::WindowMethods;
+use dom::bindings::error::{Error, ErrorResult};
+use dom::bindings::inheritance::Castable;
+use dom::bindings::js::{JS, Root};
+use dom::bindings::reflector::{DomObject, reflect_dom_object};
+use dom::bindings::str::DOMString;
+use dom::bluetooth::{AsyncBluetoothListener, Bluetooth};
+use dom::bluetooth::{canonicalize_filter, get_allowed_devices, response_async, OPTIONS_ERROR};
+use dom::bluetoothdevice::BluetoothDevice;
+use dom::globalscope::GlobalScope;
+use dom::permissionstatus::{PermissionStatus, PermissionDescriptorType};
+use dom::promise::Promise;
+use ipc_channel::ipc::{self, IpcSender};
+use js::jsapi::JSContext;
+use std::rc::Rc;
+
+const DESCRIPTOR_TYPE_ERROR: &'static str = "Wrong type of descriptor in argument list.";
+
+// https://webbluetoothcg.github.io/web-bluetooth/#bluetoothpermissionresult
+#[dom_struct]
+pub struct BluetoothPermissionResult {
+    parent: PermissionStatus,
+    devices: DOMRefCell<Vec<JS<BluetoothDevice>>>,
+}
+
+impl BluetoothPermissionResult {
+    pub fn new_inherited() -> BluetoothPermissionResult {
+        BluetoothPermissionResult {
+            parent: PermissionStatus::new_inherited(),
+            devices: DOMRefCell::new(Vec::new()),
+        }
+    }
+
+    pub fn new(global: &GlobalScope) -> Root<BluetoothPermissionResult> {
+        reflect_dom_object(box BluetoothPermissionResult::new_inherited(),
+                           global,
+                           BluetoothPermissionResultBinding::Wrap)
+    }
+
+    fn get_bluetooth(&self) -> Root<Bluetooth> {
+        self.global().as_window().Navigator().Bluetooth()
+    }
+
+    fn get_bluetooth_thread(&self) -> IpcSender<BluetoothRequest> {
+        self.global().as_window().bluetooth_thread()
+    }
+
+    // https://w3c.github.io/permissions/#create-a-permissionstatus
+    pub fn create_from_descriptor(global: &GlobalScope,
+                                  descriptor: PermissionDescriptorType)
+                                  -> Root<BluetoothPermissionResult> {
+        let bt_permission_result = BluetoothPermissionResult::new(global);
+        bt_permission_result.upcast::<PermissionStatus>().set_query(descriptor);
```

Comment:
`BluetoothPermissionResult` has a parent element which can be reached through `result.parent` or `result.upcast::<PermissionStatus>()` because it implements the [`Castable`](https://doc.servo.org/script/dom/bindings/inheritance/trait.Castable.html) trait. I use the second one here because the `parent` element is not public, and we don't have a `&self` here as argument.